man/udpsvd.8
.\" Automatically generated by Pandoc 3.1.11.1
.\"
.TH "udpsvd" "8" 
.SH NAME
udpsvd \- UDP/IP service daemon
.SH SYNOPSIS
\f[B]udpsvd\f[R] [\-hpvv] [\-u \f[I]user\f[R]] [\-l \f[I]name\f[R]] [\-i
\f[I]dir\f[R]|\-x \f[I]cdb\f[R]] [\-t \f[I]sec\f[R]] \f[I]host\f[R]
\f[I]port\f[R] \f[I]prog\f[R]
.SH DESCRIPTION
\f[B]udpsvd\f[R] creates an UDP/IP socket, binds it to the address
\f[I]host\f[R]:\f[I]port\f[R], and listens on the socket for incoming
datagrams.
.PP
If a datagram is available on the socket, \f[B]udpsvd\f[R] conditionally
starts a program, with standard input reading from the socket, and
standard output redirected to standard error, to handle this, and
possibly more datagrams.
\f[B]udpsvd\f[R] does not start the program if another program that it
has started before still is running.
If the program exits, \f[B]udpsvd\f[R] again listens to the socket until
a new datagram is available.
If there are still datagrams available on the socket, the program is
restarted immediately.
.PP
\f[B]udpsvd\f[R] optionally checks for special intructions depending on
the IP address or hostname of the client sending the datagram which not
yet was handled by a running program, see \f[B]ipsvd\-instruct\f[R](5)
for details.
.SS ATTENTION:
UDP is a connectionless protocol.
Most programs that handle user datagrams, such as \f[B]talkd\f[R](8),
keep running after receiving a datagram, and process subsequent
datagrams sent to the socket until a timeout is reached.
\f[B]udpsvd\f[R] only checks special instructions for a datagram that
causes a startup of the program; not if a program handling datagrams
already is running.
It doesn\[aq]t make much sense to restrict access through special
instructions when using such a program.
.PP
On the other hand, it makes perfectly sense with programs like
\f[B]tftpd\f[R](8), that fork to establish a separate connection to the
client when receiving the datagram.
In general it\[aq]s adequate to set up special instructions for programs
that support being run by tcpwrapper.
.SH OPTIONS
.TP
\f[I]host\f[R]
\f[I]host\f[R] either is a hostname, or a dotted\-decimal IP address, or
0.
If \f[I]host\f[R] is 0, \f[B]udpsvd\f[R] accepts datagrams to any local
IP address.
.TP
\f[I]port\f[R]
\f[B]udpsvd\f[R] accepts datagrams to \f[I]host\f[R]:\f[I]port\f[R].
\f[I]port\f[R] may be a name from /etc/services or a number.
.TP
\f[I]prog\f[R]
\f[I]prog\f[R] consists of one or more arguments.
\f[B]udpsvd\f[R] normally runs \f[I]prog\f[R] to handle a datagram, and
possibly more, that is sent to the socket, if there is no program that
was started before by \f[B]udpsvd\f[R] still running and handling
datagrams.
.TP
\f[B]\-i\f[R] \f[I]dir\f[R]
read instructions for handling new connections from the instructions
directory \f[I]dir\f[R].
See \f[B]ipsvd\-instruct\f[R](5) for details.
.TP
\f[B]\-x\f[R] \f[I]cdb\f[R]
read instructions for handling new connections from the constant
database \f[I]cdb\f[R].
The constant database normally is created from an instructions directory
by running \f[B]ipsvd\-cdb\f[R](8).
.TP
\f[B]\-t\f[R] \f[I]sec\f[R]
timeout.
This option only takes effect if the \-i option is given.
While checking the instructions directory, check the time of last access
of the file that matches the clients address or hostname if any, discard
and remove the file if it wasn\[aq]t accessed within the last
\f[I]sec\f[R] seconds; \f[B]udpsvd\f[R] does not discard or remove a
file if the user\[aq]s write permission is not set, for those files the
timeout is disabled.
Default is 0, which means that the timeout is disabled.
.TP
\f[B]\-l\f[R] \f[I]name\f[R]
local hostname.
Do not look up the local hostname in DNS, but use \f[I]name\f[R] as
hostname.
By default \f[B]udpsvd\f[R] looks up the local hostname once at startup.
.TP
\f[B]\-u\f[R] [:]\f[I]user\f[R][:\f[I]group\f[R]]
drop permissions.
Set uid and gid to the \f[I]user\f[R]\[aq]s uid and gid, as found in
\f[I]/etc/passwd\f[R], before running \f[I]prog\f[R].
If \f[I]user\f[R] is followed by a colon and a \f[I]group\f[R], set the
gid to \f[I]group\f[R]\[aq]s gid, as found in \f[I]/etc/group\f[R],
instead of \f[I]user\f[R]\[aq]s gid.
If \f[I]group\f[R] consists of a colon\-separated list of group names,
set the group ids of all listed groups.
If \f[I]user\f[R] is prefixed with a colon, the \f[I]user\f[R] and all
\f[I]group\f[R] arguments are interpreted as uid and gids respectively,
and not looked up in the password or group file.
All supplementary groups are removed.
.TP
\f[B]\-h\f[R]
Look up the client\[aq]s hostname in DNS.
.TP
\f[B]\-p\f[R]
paranoid.
After looking up the client\[aq]s hostname in DNS, look up the IP
addresses in DNS for that hostname, and forget the hostname if none of
the addresses match the client\[aq]s IP address.
You should set this option if you use hostname based instructions.
The \-p option implies the \-h option.
.TP
\f[B]\-v\f[R]
verbose.
Print verbose messages to standard output.
.TP
\f[B]\-vv\f[R]
more verbose.
Print more verbose messages to standard output.
.SH SEE ALSO
ipsvd(7), tcpsvd(8), ipsvd\-instruct(5), ipsvd\-cdb(8)
.PP
https://smarden.org/ipsvd/
.SH AUTHOR
Gerrit Pape <pape\[at]smarden.org>


package/CHANGES
  * doc/examples.html: use chpst instead of softlimit.
  * doc/install.html: how to install as non-root user.
  * doc/*.html, man/*.?: replace; converted from markdown source.
  * buffer_put.c, byte.h, byte_chr.c, byte_copy.c, byte_cr.c, byte_diff.c,
    byte_rchr.c, byte_zero.c, str_chr.c, str_len.c: sync daemontools library
    fork with runit source.

ipsvd 1.1.1
Sun, 02 Nov 2025 11:09:29 +0100
  * alloc.c, alloc_re.c, buffer_0.c, buffer_get.c, buffer_put.c, byte_chr.c,
    byte_copy.c, byte_cr.c, byte_diff.c, byte_rchr.c, byte_zero.c, error.c,
    error.h, print-cc.sh, print-ld.sh, select.h2, strerr_die.c, taia_sub.c,
    trysgprm.c, wait.h, wait_nohang.c, wait_pid.c: sync daemontools library
    fork with runit source.
  * tcpsvd.c, udpsvd.c: use parameter name "unused" in sighandler functions.
  * hasflock.h1: add #include <unistd.h>.
  * doc/*.html: !DOCTYPE.
  * doc/index.html: Related links: remove links to distributions, repair
    broken links.
  * doc/*: switch all cr.yp.to http links to https.
  * doc/*, man/*: switch all smarden.org http links to https.
  * alloc.h, buffer.c, buffer.h, buffer_get.c, buffer_put.c, buffer_write.c,
    byte.h, check-udpsvd.c, ipsvd_phcc.c, pathexec_env.c, select.h1,
    select.h2, sig.c, sig.h, sig_catch.c, socket_bind.c, socket_conn.c,
    tcpsvd.c, tcpsvd.dist, udpsvd.c, udpsvd.dist, wait.h: fix build with
    gcc15: in C23 f() means f(void), consequently signal handler functions
    now are f(int), alloc_free() and byte_zero() parameters need casts,
    buffer_unixwrite() removes "const" from parameter.
  * src/Makefile: target check depends on target default now.
  * doc/index.html: remove reference to mailing list misc@list.smarden.org.
  * doc/index.html: add reference to repository on github.com.
  * doc/install.html: put back installation step removed by mistake.

ipsvd 1.1.0
Wed, 16 Oct 2024 20:36:33 +0000
  * doc/benefits.html, doc/examples.html. doc/index.html,
    doc/install.html: remove references to sslio, sslsvd.
  * man/ipsvd-cdb.8, man/ipsvd-instruct.5, man/ipsvd.7, man/tcpsvd.8,
    man/udpsvd.8: remove references to sslio.8, sslsvd.8.
  * man/sslio.8, man/sslsvd.8: remove; obsolete.
  * tcpsvd.c, tcpsvd.dist: remove sslsvd code; obsolete.
  * *ssl*: remove; obsolete.
  * Makefile, TARGETS, check-dist, package/compile, package/upgrade:
    remove support for matrixssl; no longer build sslio, sslsvd programs.
  * doc/upgrade.html: typo.
  * tcpsvd.c, udpsvd.c: print proper group id when verbose.
  * doc/install.html: add link to sha256sum.asc.
  * doc/install.html: change mail address.
  * debian/: remove; obsolete.
  * cdb_make.c, pathexec.h, pathexec_env.c, pathexec_run.c, seek_set.c,
    sgetopt.c, sgetopt.h, subgetopt.c, subgetopt.h, tcpsvd.c, trycpp.c,
    tryflock.c, trypoll.c, trysgact.c, trysgprm.c, tryulong64.c,
    trywaitp.c, udpsvd.c, x86cpuid.c: incorporate lib changes from
    runit-2.2.0 and adapt ipsvd programs to properly build with modern
    toolchains (thx Gentoo, Debian, Void Linux).
  * prot.c, prot.h, chkshsgr.c, tryshsgr.c, hasshsgr.h?, warn-shsgr:
    remove; obsolete.
  * Makefile, TARGETS: no longer check for "shortsetgroups" (thx Leah
    Neukirchen).
  * ssl_io.c, tcpsvd.c, udpsvd.c: replace prot_uid() with setuid().

ipsvd 1.0.0
Sun, 15 Jun 2008 17:45:26 +0000

ipsvd 0.14.0
Sat, 19 Jan 2008 00:17:20 +0000
  * uidgid.c, uidgid.h: support numerical uid, gids through :uid:gid...
    (taken from the runit package).
  * tcpsvd.c, udpsvd.c, sslio.c, ssl_io.c: support numerical uid, gids,
    and colon-separated list of groups through :uid:gid... with the -u,
    -U arguments (thx Nikola Vladov).
  * man/tcpsvd.8, man/udpsvd.8, man/sslsvd.8, man/sslio.8, man/ipsvd.7:
    adapt.

ipsvd 0.13.0
Sun, 26 Aug 2007 11:04:21 +0000
  * ipsvd_check.c, man/ipsvd-instruct.5: lines starting with # are
    comments, ignore them completely.
  * ssl_io.c, man/sslio.8: support env variable SSLIO_BAD_CERTIFICATE to
    optionally accept 'bad certificate's (thx Nikola Vladov).
  * ssl_io.c, man/sslio.8, man/sslsvd.8: support env variable
    SSL_HANDSHAKE_TIMEOUT to disconnect a client if the ssl handshake
    isn't completed within some number of seconds (thx Nikola Vladov);
    add ENVIRONMENT section to man/sslsvd.8.
  * Makefile: build with matrixssl-1-8-3-open.

ipsvd 0.12.1
Sat, 04 Feb 2006 19:14:23 +0000
  * Makefile: fix build failure with local matrixssl library (thx Charlie
    Brady).

ipsvd 0.12.0
Fri, 03 Feb 2006 20:26:27 +0000
  * ssl_io.c, ssl_io.h: new; split off most code from sslio.c.
  * sslio.c: main() function only, uses ssl_io() from ssl_io.
  * tcpsvd.c: be sslsvd using ssl_io() #ifdef SSLSVD.
  * sslsvd.c: new; #define SSLSVD, #include tcpsvd.c.
  * man/sslsvd.8: new; man page for sslsvd(8).
  * man/ipsvd-cdb.8, man/ipsvd-instruct.5, man/ipsvd.7, man/sslio.8,
    man/tcpsvd.8, man/udpsvd.8: add SEE ALSO sslsvd(8).
  * sslsvd.check, sslsvd.dist: selftest for sslsvd not yet implemented.
  * doc/index.html: mention sslsvd.

ipsvd 0.11.1
Tue, 18 Oct 2005 19:02:18 +0000
  * sslio.check: tweak sslio program check (fixes testing failure on
    Debian m68k)
  * ipsvd_log.c: minor
  * Makefile, doc/install.html: adapt to matrixssl-1-7-1-open.tar.gz.

ipsvd 0.11.0
Sun, 20 Feb 2005 16:13:38 +0000
  * man/tcpsvd.8: typo (thx Bastian Kleineidam).
  * sslio.check, sslio.dist, sslio.pem: new; ssl program check.
  * Makefile: don't limit sslio program to linux systype; run ssl program
    check; minor
  * package/compile, package/upgrade: adapt.
  * doc/benefits.html, doc/index.html, doc/install.html: sslio works on
    MacOSX; adapt.

ipsvd 0.10.1
Sun, 16 Jan 2005 10:34:19 +0000
  * sslerror_str.c: fix build failure due to broken header include (thx
    Charlie Brady)
  * Makefile: target sslerror_str.o depends on matrixSsl.h.
  * sslio.c: minor.

ipsvd 0.10.0
Sat, 15 Jan 2005 11:47:55 +0000
  * doc/install.html: typo.
  * sslerror_str.c, sslerror_str.h: new; ssl error strings.

src/udpsvd.c
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <unistd.h>
#include <grp.h>
#include "dns.h"
#include "socket.h"
#include "ip4.h"
#include "ipsvd_log.h"
#include "ipsvd_check.h"
#include "ipsvd_fmt.h"
#include "ipsvd_hostname.h"
#include "ipsvd_scan.h"
#include "uidgid.h"
#include "sgetopt.h"
#include "sig.h"
#include "str.h"
#include "fmt.h"
#include "error.h"
#include "strerr.h"
#include "ndelay.h"
#include "scan.h"
#include "fd.h"
#include "wait.h"
#include "pathexec.h"

#define USAGE " [-hpv] [-u user] [-l name] [-i dir|-x cdb] [-t sec] host port prog"
#define VERSION "$Id$"

#define FATAL "udpsvd: fatal: "
#define WARNING "udpsvd: warning: "
#define INFO "udpsvd: info: "
#define DROP "udpsvd: drop: "

const char *progname;

char **prog;
const char *instructs =0;
unsigned int iscdb =0;
unsigned int verbose =0;
unsigned int lookuphost =0;
unsigned int paranoid =0;
unsigned long timeout =0;

static stralloc local_hostname ={0};
char local_ip[IP4_FMT];
char *local_port;
static stralloc remote_hostname ={0};
char remote_ip[IP4_FMT];
char remote_port[FMT_ULONG];
struct uidgid ugid;

static char seed[128];
struct sockaddr_in socka;
int socka_size =sizeof(socka);
char bufnum[FMT_ULONG];
int s;

static stralloc sa ={0};
static stralloc ips ={0};
static stralloc fqdn ={0};
static stralloc inst ={0};
static stralloc match ={0};

void usage() { strerr_die4x(111, "usage: ", progname, USAGE, "\n"); }
void die_nomem() { strerr_die2x(111, FATAL, "out of memory."); }
void fatal(char *m0) { strerr_die3sys(111, FATAL, m0, ": "); };
void fatal2(char *m0, char *m1) {
  strerr_die5sys(111, FATAL, m0, ": ", m1, ": ");
}
void warn(char *m0) { strerr_warn3(WARNING, m0, ": ", &strerr_sys); }
void warn2(char *m0, char *m1) {
  strerr_warn5(WARNING, m0, ": ", m1, ": ", &strerr_sys);
}
void drop_nomem() { strerr_die2x(111, DROP, "out of memory."); }
void drop(char *m0) { strerr_die3sys(111, DROP, m0, ": "); }
void discard(char *m0, char *m1) {
  recv(s, 0, 0, 0);
  strerr_die6sys(111, DROP, "discard data: ", m0, ": ", m1, ": ");
}

void sig_term_handler(int unused) {
  if (verbose) {
    out(INFO); flush("sigterm received, exit.\n");
  }
  _exit(0);
}

void connection_accept(int c) {
  int ac;
  char **run;
  char *args[4];
  char *ip =(char*)&socka.sin_addr;

  remote_ip[ipsvd_fmt_ip(remote_ip, ip)] =0;
  if (verbose) {
    out(INFO); out("pid ");
    bufnum[fmt_ulong(bufnum, getpid())] =0;
    out(bufnum); out(" from "); outfix(remote_ip); flush("\n");
  }
  remote_port[ipsvd_fmt_port(remote_port, (char*)&socka.sin_port)] =0;
  if (lookuphost) {
    if (ipsvd_hostname(&remote_hostname, ip, paranoid) == -1)
      warn2("unable to look up hostname", remote_ip);
    if (! stralloc_0(&remote_hostname)) drop_nomem();
  }

  if (instructs) {
    ac =ipsvd_check(iscdb, &inst, &match, (char*)instructs,
                    remote_ip, remote_hostname.s, timeout);
    if (ac == -1) discard("unable to check inst", remote_ip);
    if (ac == IPSVD_ERR) discard("unable to read", (char*)instructs);
  }
  else ac =IPSVD_DEFAULT;

  if (verbose) {
    out(INFO);
    switch(ac) {
    case IPSVD_DENY: out("deny "); break;
    case IPSVD_DEFAULT: case IPSVD_INSTRUCT: out("start "); break;
    case IPSVD_EXEC: out("exec "); break;
    }
    bufnum[fmt_ulong(bufnum, getpid())] =0;
    out(bufnum); out(" ");
    outfix(local_hostname.s); out(":"); out(local_ip);
    out(" :"); outfix(remote_hostname.s); out(":");
    outfix(remote_ip); out(":"); outfix(remote_port);
    if (instructs) {
      out(" ");
      if (iscdb) {
        out((char*)instructs); out("/");
      }
      outfix(match.s);
      if(inst.s && inst.len && (verbose > 1)) {
        out(": "); outinst(&inst);
      }
    }
    flush("\n");
  }

  if (ac == IPSVD_DENY) {
    recv(s, 0, 0, 0);
    _exit(100);
  }
  if (ac == IPSVD_EXEC) {
    args[0] ="/bin/sh"; args[1] ="-c"; args[2] =inst.s; args[3] =0;
    run =args;
  }
  else run =prog;

  if ((fd_move(0, c) == -1) || (fd_copy(1, 2) == -1))
    drop("unable to set filedescriptor");
  sig_uncatch(sig_term);
  sig_uncatch(sig_pipe);
  pathexec(run);

  discard("unable to run", (char*)*prog);
}

int main(int argc, char **argv, char *const *envp) {
  int opt;
  char *user =0;
  char *host;
  unsigned long port;
  int pid;
  int wstat;
  iopause_fd io[1];
  struct taia now;
  struct taia deadline;

  progname =*argv;

  while ((opt =getopt(argc, argv, "vu:l:hpi:x:t:V")) != opteof) {
    switch(opt) {
    case 'v':
      ++verbose;
      break;
    case 'u':
      user =(char*)optarg;
      break;
    case 'l':
      if (! stralloc_copys(&local_hostname, optarg)) die_nomem();
      if (! stralloc_0(&local_hostname)) die_nomem();
      break;
    case 'h':
      lookuphost =1;
      break;
    case 'p':
      lookuphost =1;
      paranoid =1;
      break;
    case 'i':
      if (instructs) usage();
      instructs =optarg;
      break;
    case 'x':
      if (instructs) usage();
      instructs =optarg;
      iscdb =1;
      break;
    case 't':
      scan_ulong(optarg, &timeout);
      break;
    case 'V':
      strerr_warn1(VERSION, 0);
    case '?':
      usage();
    }
  }
  argv +=optind;

  if (! argv || ! *argv) usage();
  host =(char*)*argv++;
  if (! argv || ! *argv) usage();
  local_port =(char*)*argv++;
  if (! argv || ! *argv) usage();
  prog =argv;

  if (user)
    if (! uidgids_get(&ugid, user))
      strerr_die3x(100, FATAL, "unknown user/group: ", user);

  dns_random_init(seed);
  sig_catch(sig_term, sig_term_handler);
  sig_ignore(sig_pipe);

  if (str_equal(host, "")) host ="0.0.0.0";
  if (str_equal(host, "0")) host ="0.0.0.0";

  if (! ipsvd_scan_port(local_port, "udp", &port))
    strerr_die3x(100, FATAL, "unknown port number or name: ", local_port);

  if (! stralloc_copys(&sa, host)) die_nomem();
  if ((dns_ip4(&ips, &sa) == -1) || (ips.len < 4))
    if (dns_ip4_qualify(&ips, &fqdn, &sa) == -1)
      fatal2("unable to look up ip address", host);
  if (ips.len < 4)
    strerr_die3x(100, FATAL, "unable to look up ip address: ", host);
  ips.len =4;
  if (! stralloc_0(&ips)) die_nomem();
  local_ip[ipsvd_fmt_ip(local_ip, ips.s)] =0;
  if (! local_hostname.len) {
    if (dns_name4(&local_hostname, ips.s) == -1)
      fatal("unable to look up local hostname");
    if (! stralloc_0(&local_hostname)) die_nomem();
  }

  if (! lookuphost) {
    if (! stralloc_copys(&remote_hostname, "")) die_nomem();
    if (! stralloc_0(&remote_hostname)) die_nomem();
  }

  if ((s =socket_udp()) == -1) fatal("unable to create socket");
  if (socket_bind4_reuse(s, ips.s, port) == -1)
    fatal("unable to bind socket");
  ndelay_off(s);

  if (user) { /* drop permissions */
    if (setgroups(ugid.gids, ugid.gid) == -1) fatal("unable to set groups");
    if (setgid(*ugid.gid) == -1) fatal("unable to set gid");
    if (setuid(ugid.uid) == -1) fatal("unable to set uid");
  }
  close(0);

  if (verbose) {
    out(INFO); out("listening on "); outfix(local_ip); out(":");
    outfix(local_port);
    if (user) {
      bufnum[fmt_ulong(bufnum, ugid.uid)] =0;
      out(", uid "); out(bufnum);
      bufnum[fmt_ulong(bufnum, *ugid.gid)] =0;
      out(", gid "); out(bufnum);
    }
    flush(", starting.\n");
  }

  io[0].fd =s;
  io[0].events =IOPAUSE_READ;
  io[0].revents =0;
  for (;;) {
    taia_now(&now);
    taia_uint(&deadline, 3600);
    taia_add(&deadline, &now, &deadline);
    iopause(io, 1, &deadline, &now);

    if (io[0].revents | IOPAUSE_READ) {
      io[0].revents =0;
      while ((pid =fork()) == -1) {
        warn("unable to fork, sleeping");
        sleep(5);
      }
      if (pid == 0) { /* child */
        if (recvfrom(s, 0, 0, MSG_PEEK, (struct sockaddr *)&socka,
                     &socka_size) == -1) drop("unable to read from socket");
        connection_accept(s);
      }
      while (wait_pid(&wstat, pid) == -1) warn("error waiting for child");
      if (verbose) {
        out(INFO); out("end ");
        bufnum[fmt_ulong(bufnum, pid)] =0;
        out(bufnum); flush("\n");
      }
    }
  }
  _exit(0);
}
